### ISSUE TITLE
coarse grid solver w. AMG uses system_matrix.vmult() 

### ISSUE THREAD

**Issue by richardschu** [exadg/exadg#691]
Even pre #681, we used the `coarse_operator.system_matrix` for the Krylov solver on the coarse level.
See [here](https://github.com/exadg/exadg/pull/681/files#r1702718208).

with #681, `precondition_amg.h` also includes [`PreconditionerAMG::apply_krylov_solver_with_amg_preconditioner()`](https://github.com/exadg/exadg/blob/b60088eddddb04fe6ddfa7f54a9512dd0fa11e0b/include/exadg/solvers_and_preconditioners/preconditioners/preconditioner_amg.h#L393C3-L393C46) 
which is only used in [MGCoarseKrylov::operator()(dst,src)](https://github.com/exadg/exadg/blob/b60088eddddb04fe6ddfa7f54a9512dd0fa11e0b/include/exadg/solvers_and_preconditioners/multigrid/coarse_grid_solvers.h#L169)

With this setup, the type conversion coarsest level: double ---> double (in case of Trilinos::ML) ---> float
happens explicitly when we solve, once, using [`apply_function_in_double_precision()`](https://github.com/exadg/exadg/blob/b60088eddddb04fe6ddfa7f54a9512dd0fa11e0b/include/exadg/solvers_and_preconditioners/preconditioners/preconditioner_amg.h#L418-L426).

However, we could also just use `underlying_opartor` directly as shown in #691.
This way, we would convert single--->double--->single implicitly all the time in each iteration (which are hopefully <30 ?).

So there could be performance loss.
However, I was having trouble using the current implementation using `preconditioner_amg.apply_krylov_solver_with_amg_preconditioner()` which just gave me NaN after one iteration, but *not* in the first application of the solver. This is still a mistery to me, especially since it does not happen on two other machines, only on Fritz.
I will only investigate this issue further, if we decide to keep `PreconditionerAMG::apply_krylov_solver_with_amg_preconditioner()`. Is it a-priori clear, that type conversion is better/worse than `system_matrix.vmult()` vs. the matrix-free `operator.vmult()`? What am I not seeing? #681 specifically introduced this functionality to use it only in a single place, so there is a reason for that, which I do not see yet.

Thanks for any comments in advance :+1:

**Comment by nfehn** [comment]
Overall, I do not understand where this behavior comes from. Currently, I do not see that the refactoring I did (shifting code from the coarse-grid solver implementation to the AMG preconditioner) changes the code. However, what I changed intentionally in #681 is the replacement of the vector `src` by the vector `r` in the coarse-grid solver, see issue https://github.com/exadg/exadg/issues/682. @richardschu did you check that the (wrong) behavior you observe right now does not originate from this change?

> At the same time, AMG alone (not used as preconditioner for a Krylov method) on the coarse grid is totally fine, MGCoarseKrylov with PointJacobi preconditioner is also fine. The problem is hence definitely in 'apply_krylov_solver_with_amg_preconditioner()'.

I am not sure if this conclusion is necessarily correct. I think the change `src` -> `r` could also be a potential explanation, because this change in #681 only affects coarse-grid solvers of type "Krylov+AMG preconditioner" if I remember correctly. This seems to be the case that you now observe to no longer work as before the changes in PR #681.

**Comment by nfehn** [comment]
As described here https://github.com/exadg/exadg/pull/692#issuecomment-2268642765, the old code has run the whole Krylov coarse-grid solver in double precision according to my understanding. In my opinion, we should first discuss why we decided to do it like this or why this is needed before we think about changing the code, e.g. as suggested in #692. @kronbichler @peterrum do you remember why we run the coarse-grid Krylov solver in double precision in case of AMG preconditioners?

**Comment by peterrum** [comment]
> do you remember why we run the coarse-grid Krylov solver in double precision in case of AMG preconditioners?

So that we don't have to copy the vectors from double to float in every coarse-grid iteration.

**Comment by richardschu** [comment]
> > do you remember why we run the coarse-grid Krylov solver in double precision in case of AMG preconditioners?
> 
> So that we don't have to copy the vectors from double to float in every coarse-grid iteration.

makes sense. but is this always preferrable over using the matrix-free operator? We compare matrix-free operator + conversion in every iteration VS system_matrix.vmult() + stay in double throughout the Krylov solver.

**Comment by richardschu** [comment]
> Overall, I do not understand where this behavior comes from. Currently, I do not see that the refactoring I did (shifting code from the coarse-grid solver implementation to the AMG preconditioner) changes the code. However, what I changed intentionally in #681 is the replacement of the vector `src` by the vector `r` in the coarse-grid solver, see issue #682. @richardschu did you check that the (wrong) behavior you observe right now does not originate from this change?
> 
> > At the same time, AMG alone (not used as preconditioner for a Krylov method) on the coarse grid is totally fine, MGCoarseKrylov with PointJacobi preconditioner is also fine. The problem is hence definitely in 'apply_krylov_solver_with_amg_preconditioner()'.
> 
> I am not sure if this conclusion is necessarily correct. I think the change `src` -> `r` could also be a potential explanation, because this change in #681 only affects coarse-grid solvers of type "Krylov+AMG preconditioner" if I remember correctly. This seems to be the case that you now observe to no longer work as before the changes in PR #681.

Thanks for your comment.
I played around with it (since it was once of the changes of #681 as you said), but I do not have a reliable track of these tests. From what I remember, it was not the decisive ingredient. Does `operator_is_singular` evaluate to true on the MG levels or are we talking about pressure spaces where it is `true`?
